[UBSan][BoundsSafety] Implement support for more expressive "trap reasons"

delcypher opened the pull request:
In 29992cfd628ed5b968ccb73b17ed0521382ba317 (#145967) support was added for "trap reasons" on traps emitted in UBSan in trapping mode (e.g. `-fsanitize-trap=undefined`). This improved the debugging experience by attaching the reason for trapping as a string on the debug info on trap instructions. Consumers such as LLDB can display this trap reason string when the trap is reached.

A limitation of that patch is that the trap reason string is hard-coded for each `SanitizerKind` even though the compiler actually has much more information about the trap available at compile time that could be shown to the user.

This patch is an incremental step in fixing that. It consists of two main steps.

**1. Introduce infrastructure for building trap reason strings**

To make it convenient to construct trap reason strings this patch re-uses Clang's powerful diagnostic infrastructure to provide a convenient API for constructing trap reason strings. This is achieved by:

* Introducing a new `Trap` diagnostic kind to represent trap diagnostics in TableGen files.
* Adding a new `Trap` diagnostic component. While this part probably isn't technically necessary it seemed like I should follow the existing convention used by the diagnostic system.
* Adding `DiagnosticTrapKinds.td` to describe the different trap reasons.
* Add the `TrapReasonBuilder` and `TrapReason` classes to provide an interface for constructing trap reason strings and the trap category. Note this API while similar to `DiagnosticBuilder` has different semantics which are described in the code comments. In particular the behavior when the destructor is called is very different.
* Adding `CodeGenModule::BuildTrapReason()` as a convenient constructor for the `TrapReasonBuilder`.

This use of the diagnostic system is a little unusual in that the emitted trap diagnostics aren't actually consumed by normal diagnostic consumers (e.g. the console). Instead the `TrapReasonBuilder` is just used to format a string, so in effect the builder is somewhat analagous to "printf". However, re-using the diagnostics system in this way brings a several benefits:

* The powerful diagnostic templating languge (e.g. `%select`) can be used.
* Formatting Clang data types (e.g. `Type`, `Expr`, etc.) just work out-of-the-box.
* Describing trap reasons in tablegen files opens the door for translation to different languages in the future.
* The `TrapReasonBuilder` API is very similar to `DiagnosticBuilder` which makes it easy to use by anyone already familiar with Clang's diagnostic system.

While UBSan is the first consumer of this new infrastructure the intent is to use this to overhaul how trap reasons are implemented in the `-fbounds-safety` implementation (currently exists downstream).

**2. Apply the new infrastructure to UBSan checks for arithmetic overflow**

To demonstrate using `TrapReasonBuilder` this patch applies it to UBSan traps for arithmetic overflow. The intention is that we would iteratively switch to using the `TrapReasonBuilder` for all UBSan traps where it makes sense in future patches.

Previously for code like

```
int test(int a, int b) { return a + b; }
```

The trap reason string looked like

```
Undefined Behavior Sanitizer: Integer addition overflowed
```

now the trap message looks like:

```
Undefined Behavior Sanitizer: signed integer addition overflow in 'a + b'
```

This string is much more specific because

* It explains if signed or unsigned overflow occurred
* It actually shows the expression that overflowed

One possible downside of this approach is it may blow up Debug info size because now there can be many more distinct trap reason strings. To allow users to avoid this a new driver/cc1 flag `-fsanitize-debug-trap-reasons=` has been added which can either be `none` (disable trap reasons entirely), `basic` (use the per `SanitizerKind` hard coded strings), and `detailed` (use the new expressive trap reasons implemented in this patch). The default is `detailed` to give the best out-of-the-box debugging experience. The existing `-fsanitize-debug-trap-reasons` and  `-fno-sanitize-debug-trap-reasons` have been kept for compatibility and are aliases of the new flag with `detailed` and `none` arguments passed respectively.


rdar://158612755

delcypher committed:
[UBSan][BoundsSafety] Implement support for more expressive "trap reasons"

In 29992cfd628ed5b968ccb73b17ed0521382ba317 (#145967) support was added
for "trap reasons" on traps emitted in UBSan in trapping mode (e.g.
`-fsanitize-trap=undefined`). This improved the debugging experience by
attaching the reason for trapping as a string on the debug info on trap
instructions. Consumers such as LLDB can display this trap reason string
when the trap is reached.

A limitation of that patch is that the trap reason string is hard-coded
for each `SanitizerKind` even though the compiler actually has much more
information about the trap available at compile time that could be shown
to the user.

This patch is an incremental step in fixing that. It consists of two
main steps.

**1. Introduce infrastructure for building trap reason strings **

To make it convenient to construct trap reason strings this patch
re-uses Clang's powerful diagnostic infrastructure to provide a
convenient API for constructing trap reason strings. This is achieved
by:

* Introducing a new `Trap` diagnostic kind to represent trap
  diagnostics in TableGen files.
* Adding a new `CodeGen` diagnostic component. While this part probably
  isn't technically necessary it seemed like I should follow the
  existing convention used by the diagnostic system.
* Adding `DiagnosticCodeGenKinds.td` to describe the different trap
  reasons.
* Add the `RuntimeTrapDiagnosticBuilder` class to provide an interface
  for constructing trap reason strings and the trap category. Note this
  API while similar to `DiagnosticBuilder` has different semantics which
  are described in the code comments. In particular the behavior when
  the destructor is called is very different.
* Adding `CodeGenModule::RuntimeDiag()` as a convenient constructor for
  the `RuntimeTrapDiagnosticBuilder`.

This use of the diagnostic system is a little unusual in that the
emitted trap diagnostics aren't actually consumed by normal diagnostic
consumers (e.g. the console). Instead the `RuntimeTrapDiagnosticBuilder`
is just used to format a string, so in effect the builder is somewhat
analagous to "printf". However, re-using the diagnostics system in this
way brings a several benefits:

* The powerful diagnostic templating languge (e.g. `%select`) can be
  used.
* Formatting Clang data types (e.g. `Type`, `Expr`, etc.) just
  work out-of-the-box.
* Describing trap reasons in tablegen files opens the door for
  translation to different languages in the future.
* The `RuntimeTrapDiagnosticBuilder` API is very similar to
  `DiagnosticBuilder` which makes it easy to use by anyone already
  familiar with Clang's diagnostic system.

While UBSan is the first consumer of this new infrastructure the intent
is to use this to overhaul how trap reasons are implemented in the
`-fbounds-safety` implementation (currently exists downstream).

**2. Apply the new infrastructure to UBSan checks for arithmetic overflow**

To demonstrate using `RuntimeTrapDiagnosticBuilder` this patch applies
it to UBSan traps for arithmetic overflow. The intention is that we
would iteratively switch to using the `RuntimeTrapDiagnosticBuilder` for
all UBSan traps where it makes sense in future patches.

Previously for code like

```
int test(int a, int b) { return a + b; }
```

The trap reason string looked like

```
Undefined Behavior Sanitizer: Integer addition overflowed
```

now the trap message looks like:

```
Undefined Behavior Sanitizer: signed integer addition overflow in 'a + b'
```

This string is much more specific because

* It explains if signed or unsigned overflow occurred
* It actually shows the expression that overflowed

This seems a lot more helpful.

One possible downside of this approach is it may blow up Debug info size
because now there can be many more distinct trap reason strings. If this
is a concern we may want to add a flag to make it possible to continue
to use the original hard-coded trap messages to avoid increasing the
size of Debug info.

rdar://158612755

delcypher commented:
@thurstond 
> From example.c:4:35, we can surmise the error is return a + b. Are there any common cases where lldb wouldn't have the debug info to be able to print the source code, but would have trap reasons debug info; or where having the trap reason would offer a substantial benefit over needing to look at the source code (which I assume is the logical next step for debugging even if we had the detailed trap reason)?

I suspect that this would be helpful with anything involving macros because the expression printed in the trap reason will be the fully expanded expression, whereas looking at the source code will just show you the unexpanded macros.

To be clear printing the Expr that overflowed is not a critical part of this PR and I'm happy to drop it if we do not think its helpful. I will be using the ability to print Clang expressions in trap messages in the implementation of `-fbounds-safety` but UBSan does not need to do this. In this PR I'm trying to show the kind of thing that can be done when we have `RuntimeTrapDiagnosticBuilder` at our disposal.

> Could the same benefit be obtained cheaply by separating handlers (e.g., AddOverflow -> SignedAddOverflow, UnsignedAddOverflow)?

For this **specific** example technically yes but it gets unruly very quickly as you try to handle more cases. You'd have to do the same for subtraction, and multiplication too. 

* My GSoC student (@anthonyhatran ) is looking for places where the compiler knows more information than is currently in the UBSan trap message and has identified several places where we could improve this (see https://github.com/llvm/llvm-project/pull/153845). Also, I am extremely wary of modifying `SanitizerKind` because the enum values are encoded in trap instructions for `x86_64` and `arm64` which has effectively become "ABI" which we should be very careful to not break.
* We want to improve the trap messages on `-fbounds-safety` traps and I want to use this infrastructure to do that. Although I could've implemented `RuntimeTrapDiagnosticBuilder` downstream I'd prefer to implement this upstream so that UBSan can benefit from it and to avoid merge conflicts.

Sirraide commented on clang/include/clang/Basic/DiagnosticCodeGenKinds.td:
```diff
@@ -0,0 +1,26 @@
+let CategoryName = "Undefined Behavior Sanitizer" in {
+
+def trap_ubsan_arith_overflow : Trap<
+  "%select{unsigned|signed}0 integer "
+  "%select{addition|subtraction|multiplication}1 overflow in %2">;
```
Can we use `%enum_select` for these?

delcypher replied:
I didn't even know that existed. I try using it.

delcypher replied:
Switched to `%enum_select` :)

Sirraide commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -1824,6 +1825,12 @@ class CodeGenModule : public CodeGenTypeCache {
     return PAlign;
   }
 
+  /// Helper function to construct a RuntimeTrapDiagnosticBuilder
+  [[nodiscard]] std::unique_ptr<RuntimeTrapDiagnosticBuilder>
+  RuntimeDiag(unsigned DiagID) {
+    return std::make_unique<RuntimeTrapDiagnosticBuilder>(&getDiags(), DiagID);
+  }
+
```
We don’t usually heap-allocate these; is there any reason to not just return a `RuntimeTrapDiagnosticBuilder` directly?

Sirraide replied:
So if I’m understanding this correctly, the reason why you can’t ‘just put it on the stack’ is because the act of creating the diagnostic builder also eventually formats the diagnostic (because that’s the whole point of it after all), and we only want to create the diagnostic if we actually need it. That makes sense, and having a separate class to hold the message would probably work. Another idea I just had is to use `std::optional`:
```
std::optional<RuntimeTrapDiagnosticBuilder> DB;
if (some_condition) {
    DB = CGM.RuntimeDiag(trap_ubsan_arith) << 0 << someExpr;
};
consume(DB);
```
I wonder if this avoids copying the string after we format if we can just hold on to the diagnostic builder; we’d even avoid formatting the diagnostic until it’s actually used (in case we ever decide to create but not use it), and we’d be able to add more diagnostic arguments to it later on—that said, I’m not sure how well this approach would work, so feel free to go with whatever approach you think makes this easier to implement and maintain. ;Þ

Sirraide reviewed (commented):
Reusing the existing diagnostics machinery for this sgtm.

delcypher commented on clang/include/clang/Basic/AllDiagnosticKinds.inc:
```diff
@@ -30,4 +30,5 @@
 #include "clang/Basic/DiagnosticAnalysisKinds.inc"
 #include "clang/Basic/DiagnosticRefactoringKinds.inc"
 #include "clang/Basic/DiagnosticInstallAPIKinds.inc"
+#include "clang/Basic/DiagnosticCodeGenKinds.inc"
```
So the reason I picked the name `CodeGen` is because it seemed like the Diagnostics subsystem named the component after the library it was used in (e.g. `Analysis`, `InstallAPI`, `Sema`, etc...). An outlier to that I suppose would be the `Common` component. If you are ok with breaking (what looks like) convention then I'm happy with `DiagnosticTrapKinds` because it's a much better name.

Sirraide replied:
> So the reason I picked the name CodeGen is because it seemed like the Diagnostics subsystem named the component after the library it was used in (e.g. Analysis, InstallAPI, Sema, etc...)

I figured that was just a coincidence but I don’t actually know the history behind it; but either way, I personally would argue we should go with sth like `DiagnosticTrapKinds` regardless.

delcypher commented:
> Reusing the existing diagnostics machinery for this sgtm.

Thank you for being open to this. I was a little worried that using the diagnostics machinery for this would be considered in appropriate.

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3721,7 +3722,8 @@ static void emitCheckHandlerCall(CodeGenFunction &CGF,
     SanitizerHandler CheckHandler, ArrayRef<llvm::Constant *> StaticArgs,
-    ArrayRef<llvm::Value *> DynamicArgs) {
+    ArrayRef<llvm::Value *> DynamicArgs,
+    std::unique_ptr<RuntimeTrapDiagnosticBuilder> RTDB) {
```
I did that because I thought of the function as "taking ownership" of the memory so that it is immediately released when `CodeGenFunction::EmitCheck` returns. It's not super critical though. If we end up using my alternative proposal (https://github.com/llvm/llvm-project/pull/154618#discussion_r2292181048) to avoid heap allocation then this will go away entirely.

ilovepi commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1535,72 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+  RuntimeTrapDiagnosticBuilder(const RuntimeTrapDiagnosticBuilder &) = delete;
+  RuntimeTrapDiagnosticBuilder(const RuntimeTrapDiagnosticBuilder &&) = delete;
+
+  /// \return Format the trap message and return it. Note the lifetime of
+  /// the underlying storage pointed to by the returned StringRef is the same
+  /// as the lifetime of this class. This means it is likely unsafe to store
+  /// the returned StringRef.
+  StringRef getMessage();
```
This sounds like a bit of a footgun... should we just return a string in this case?  Same for `getCategory()`

delcypher replied:
Yeah it is. I tried to partially mitigate it by having `TrapDiagnosticBuilder` always be heap allocated and managed by unique pointer.

I have an alternative design proposal in this comment https://github.com/llvm/llvm-project/pull/154618#discussion_r2292181048 that would solve this too because it moves storage of the formatted string outside of the `TrapDiagnosticBuilder`.

ilovepi replied:
The alternative API sounds reasonable. Stack allocation is kind of free in that we just bump a pointer.

ilovepi commented on clang/include/clang/Basic/DiagnosticCodeGen.h:
```diff
@@ -0,0 +1,15 @@
+//===--- DiagnosticCodeGen.h - Diagnostics for libcodegen -------*- C++ -*-===//
```
nit: We've updated the style guide to no longer include the file names in the header block.
https://llvm.org/docs/CodingStandards.html#file-headers

delcypher replied:
Thanks for the heads up. I just copied an existing header not realizing the style guide had changed.

delcypher commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1535,72 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+  StringRef getCategory();
+
+private:
+  llvm::SmallVector<char, 64> MessageStorage;
```
So I originally did this because I needed to call `Diagnostic::FormatDiagnostic(SmallVectorImpl<char> &OutStr)`. However, I didn't realize the `SmallString` sub-classes `SmallVector` so this change should probably just work.

delcypher replied:
Fixed

delcypher committed:
Add `[[nodiscard]]` to `RuntimeTrapDiagnosticBuilder`

delcypher committed:
Remove unnecessary incldue

delcypher committed:
Use modern header block for `DiagnosticTrap.h` and add a little bit of
documentation.

delcypher commented on clang/include/clang/Basic/DiagnosticCodeGen.h:
```diff
@@ -0,0 +1,15 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
Done

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -33,6 +33,7 @@
 #include "clang/AST/StmtVisitor.h"
 #include "clang/Basic/Builtins.h"
 #include "clang/Basic/CodeGenOptions.h"
+#include "clang/Basic/DiagnosticParse.h"
```
Fixed.

delcypher committed:
Fix some formatting

delcypher commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1535,72 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+/// * Users of this class that want to retrieve the "trap reason" should call
+///   call the `getMessage()` and `getCategory()` and use those results before
+///   the builder is destroyed.
+/// * Unlike the `DiagnosticBuilder` the `RuntimeDiagnosticBuilder` should never
+/// be created as a temporary (i.e. rvalue) and instead should be stored. This
```
Added the attribute

delcypher committed:
Use %enum_select

delcypher committed:
Use llvm::SmallString instead of llvm::SmallVector

delcypher committed:
Move `RuntimeTrapDiagnosticBuilder` to a safer API where the storage and
transport of the "Trap Reason" is moved out of
`RuntimeTrapDiagnosticBuilder` into a new helper class called
`TrapReason`.

delcypher committed:
Fix `tools/clang/test/Misc/warning-flags.c` by teaching `diagtool` to
not treat trap diagnostics as a warning diagnostic.

delcypher commented:
@Sirraide @ilovepi I've changed the implementation to use a `TrapReason` object to store the trap message and category. This allows cleaning up the `TrapDiagnosticBuilder` API (maybe it should be called `TrapReasonBuilder` now?)  so it's much more similar to the `DiagnosticBuilder` and much less of a foot-gun.

Let me know what you think.

Sirraide commented on clang/lib/Basic/DiagnosticIDs.cpp:
```diff
@@ -391,6 +395,8 @@ unsigned DiagnosticIDs::getCustomDiagID(CustomDiagDesc Diag) {
 }
 
 bool DiagnosticIDs::isWarningOrExtension(unsigned DiagID) const {
+  // FIXME: Returning `true` for notes, remarks, traps, and
```
Yeah, either the implementation or the name of this function is just wrong; I suppose what should be done here is to look at how it’s used and then either rename it or change what it does, but yeah, that should be a separate patch.

Sirraide reviewed (commented):
Just some minor comments; the overall approach looks fine to me.

delcypher commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1536,103 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+  /// \return the trap category (e.g. "Undefined Behavior Sanitizer")
+  StringRef getCategory() const { return Category; }
+
+  bool isEmpty() const { return Message.size() == 0 && Category.size() == 0; }
```
Ah. So there is a reason the logic is written this way. This is because in the `-fbounds-safety` implementation downstream we sometimes emit a category with an empty message. We do this for cases in codegen where (unfortunately) a bunch of the necessary information to explain the trap has been lost and so we emit the trap with a category so its clear what type of trap it is but keep the message empty because there's nothing useful that can be said about the specific reason for trapping.

Sirraide replied:
Ah, that makes sense; it might be helpful to add a comment explaining that perhaps.

delcypher commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1536,103 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+/// \endcode
+///
+///
+class RuntimeTrapDiagnosticBuilder : public DiagnosticBuilder {
```
I think the classes will only ever be used in CodeGen. I'll try moving them.

delcypher replied:
I figured it out. I forgot that when I moved the class I also moved it into a different namespace so I forgot to update the forward declaration in `Diagnostic.h` to be in the correct namespace. So doing this "works":

```
index af26a04d9488..b957ca331781 100644
--- a/clang/include/clang/Basic/Diagnostic.h
+++ b/clang/include/clang/Basic/Diagnostic.h
@@ -60,6 +60,9 @@ class Preprocessor;
 class SourceManager;
 class StoredDiagnostic;
 
+namespace CodeGen {
+  class TrapReasonBuilder;
+} // namespace CodeGen
 namespace tok {
 
 enum TokenKind : unsigned short;
@@ -1234,6 +1237,7 @@ class DiagnosticBuilder : public StreamingDiagnostic {
   friend class DiagnosticsEngine;
   friend class PartialDiagnostic;
   friend class Diagnostic;
+  friend class CodeGen::TrapReasonBuilder;
 
   mutable DiagnosticsEngine *DiagObj = nullptr;
 
@@ -1260,13 +1264,14 @@ class DiagnosticBuilder : public StreamingDiagnostic {
 
   DiagnosticBuilder() = default;
 
-protected:
+
   DiagnosticBuilder(DiagnosticsEngine *DiagObj, SourceLocation DiagLoc,
                     unsigned DiagID);
 
   DiagnosticsEngine *getDiagnosticsEngine() const { return DiagObj; }
   unsigned getDiagID() const { return DiagID; }
 
+protected:
   /// Clear out the current diagnostic.
   void Clear() const {
     DiagObj = nullptr;
```

I think it's very ugly though to have the `CodeGen` namespace mentioned in the `Diagnostic.h` so I'd much rather not rely on using `friend class` and instead just rely on using the new `protected` accessor methods I added.

delcypher committed:
Update clang/include/clang/Basic/Diagnostic.td

Co-authored-by: Sirraide <aeternalmail@gmail.com>

delcypher commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1536,103 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+  RuntimeTrapDiagnosticBuilder(const RuntimeTrapDiagnosticBuilder &) = delete;
+  RuntimeTrapDiagnosticBuilder(const RuntimeTrapDiagnosticBuilder &&) = delete;
+
+  /// \return Format the trap message into `Storage`.
+  void getMessage(SmallVectorImpl<char> &Storage);
+
+  /// \return Return the trap category. These are the `CategoryName` property
+  /// of `trap` diagnostics declared in `DiagnosticTrapKinds.td`.
+  StringRef getCategory();
```
Currently they don't need to be so I'll make them private.

delcypher committed:
Update clang/include/clang/Basic/DiagnosticTrap.h

Co-authored-by: Sirraide <aeternalmail@gmail.com>

delcypher committed:
Update clang/lib/CodeGen/CGExprScalar.cpp

Co-authored-by: Sirraide <aeternalmail@gmail.com>

delcypher committed:
Rename classes and move into CodeGen

delcypher committed:
Small clean ups

delcypher committed:
Drop top-level comment in header

delcypher commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -1534,6 +1536,103 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
+    if (isEmpty())
+      return nullptr;
+    return this;
+  }
```
I've removed this.

delcypher commented:
@Sirraide Thanks for all the super useful feedback. Is this good to go (aside from the UBSan changes)?

delcypher committed:
Add missing include to unbreak the build

thurstond commented:
> To be clear printing the Expr that overflowed is not a critical part of this PR and I'm happy to drop it if we do not think its helpful.

I agree it can be useful in many cases, but it could also make the debug info unbearably large. (Anecdata: Big Google disables the current trap-reasons annotations in their internal toolchain, because it made the debug info up to 15% larger, even without these expressions.)

Would the ability to control it via a flag be useful? (e.g., either convert "-fsanitize-debug-trap-reasons" into a ternary flag - something like "off"/"basic"/"detailed" - or add a separate flag.) Just my tuppence. I don't object to the current patch, even without adding the flag, since users can still disable trap reasons entirely.

delcypher committed:
Add missing `isEmpty()` test to fix tests

Sirraide commented on clang/include/clang/Basic/Diagnostic.td:
```diff
@@ -144,7 +145,8 @@ class Extension<string str> : Diagnostic<str, CLASS_EXTENSION, SEV_Ignored>;
 // Notes can provide supplementary information on errors, warnings, and remarks.
 class Note<string str>      : Diagnostic<str, CLASS_NOTE, SEV_Fatal/*ignored*/>;
-
+// Traps messages attached to traps in debug info.
```
```suggestion
// Trap messages attached to traps in debug info.
```
Typo

Sirraide commented on clang/lib/CodeGen/TrapReasonBuilder.h:
```diff
@@ -0,0 +1,112 @@
+/// It is intended that this object be allocated on the stack.
+class TrapReason {
+public:
+  TrapReason() {}
```
```suggestion
  TrapReason() = default;
```

Sirraide commented on clang/lib/CodeGen/TrapReasonBuilder.h:
```diff
@@ -0,0 +1,112 @@
+///
+/// * This class should only be used with trap diagnostics (declared in
+/// `DiagnosticTrapKinds.td`).
+/// * The `RuntimeTrapDiagnosticBuilder` does not emit diagnostics to the normal
```
```suggestion
/// * The `TrapReasonBuilder` does not emit diagnostics to the normal
```

Sirraide reviewed (commented):
Some very minor things, but other than this it lgtm; I’ll leave it to someone more familiar w/ UBSan to approve this though.

delcypher committed:
Fix inconsistent ordering

delcypher commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1839,19 +1842,43 @@ void ScalarExprEmitter::EmitBinOpCheck(
-      case BO_Mul: Check = SanitizerHandler::MulOverflow; break;
-      default: llvm_unreachable("unexpected opcode for bin op check");
+      case BO_Add: {
+        ArithOverflowKind = diag::UBSanArithKind::Add;
```
Oops. Thanks for catching that.

delcypher committed:
Update clang/include/clang/Basic/Diagnostic.td

Co-authored-by: Sirraide <aeternalmail@gmail.com>

delcypher committed:
Update clang/lib/CodeGen/TrapReasonBuilder.h

Co-authored-by: Sirraide <aeternalmail@gmail.com>

delcypher committed:
Update clang/lib/CodeGen/TrapReasonBuilder.h

Co-authored-by: Sirraide <aeternalmail@gmail.com>

delcypher committed:
Fix out-of-date comment

delcypher committed:
Make `-fsanitize-debug-trap-reasons=` a parameterized flag that takes
either `none`, `basic`, or `detailed`.

The old boolean flags are included for compatibility and are aliases of
the new flag.

The `basic` mode is the behavior that existed before this patch, and
`detailed` is the newer more expressive trap reasons.

delcypher committed:
Fix formatting

ilovepi commented on clang/test/DebugInfo/Generic/ubsan-trap-reason-add-overflow.c:
```diff
@@ -1,9 +1,32 @@
+// DETAILED: [[LOC]] = !DILocation(line: 0, scope: [[MSG:![0-9]+]], {{.+}})
+// DETAILED: [[MSG]] = distinct !DISubprogram(name: "__clang_trap_msg$Undefined Behavior Sanitizer$unsigned integer addition overflow in 'c + d'"
+
+// In "Basic" mode both the trap reason is shared by both functions.
```
```suggestion
// In "Basic" mode the trap reason is shared by both functions.
```
Was this a typo?

delcypher replied:
Yep. Thanks for catching this. I've fixed the three test cases that had this typo.

ilovepi reviewed (commented):
LGTM from my perspective. Its worth noting that I'm not one of the normal maintainers of UBSan though. My team is just eager for this feature for our embedded/kernel users. I'll defer to @vitalybuka and @thurstond since they deal with this part of the codebase much more frequently than me.

delcypher committed:
Fix typos in some test comments

vitalybuka reviewed (approved):
LGTM, but please address others comments

delcypher commented:
@vitalybuka Thanks for approving. I think I addressed everyone's comments so I'm going to rebase, check everything still works, and then land.

thurstond commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -166,7 +174,9 @@ Non-comprehensive list of changes in this release
 New Compiler Flags
 ------------------
-- New option ``-fno-sanitize-annotate-debug-info-traps`` added to disable emitting trap reasons into the debug info when compiling with trapping UBSan (e.g. ``-fsanitize-trap=undefined``).
+- New option ``-fno-sanitize-debug-trap-reasons`` added to disable emitting trap reasons into the debug info when compiling with trapping UBSan (e.g. ``-fsanitize-trap=undefined``).
```
> when compiling with trapping UBSan (e.g. ``-fsanitize-trap=undefined``).

This is somewhat misleading because by itself -fsanitize-trap=undefined doesn't enable UBSan (as  you and Anthony have noted in https://discourse.llvm.org/t/rfc-emit-a-warning-when-fsanitize-trap-is-passed-without-associated-fsanitize/87893)

delcypher committed:
Address some feedback for the release notes and also explain about
unoptimized-vs-optimized builds
